{
 "metadata": {
  "name": "",
  "signature": "sha256:b46eaee7cbdc6fdc78b450f0ff0b2fc7bd1bed01ef0a091703d9e12e5c7856cf"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "In this recipe, we're going to develop a small `Workflow` to process sequence data sourced from [RNAcentral](http://rnacentral.org/). The focus of this recipe is to highlight how the `Workflow` object can decompose complex logic into smaller blocks that are easier to verify. The structure of the Notebook is to first develop methods that allow us to query RNAcentral and NCBI, where the former is our primary source of data and the latter is for obtaining lineage details."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "First, we're going to import the `requests` library which makes querying remote RESTful interfaces very easy. More detail on `requests` can be found [here](http://docs.python-requests.org/en/latest/). Second, we're going to import the `Workflow` object and two decorators."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import requests\n",
      "from skbio.workflow import Workflow, method, requires"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Next, let's put together a method that allows us to query RNAcentral. Details about its RESTful interface can be found [here](http://rnacentral.org/api/), and is an excellent example of a well thought out web interface. This method will allow us to perform arbitrary queries, and it'll return a fair bit of detail about each sequence found."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def query_rnacentral(query, max_records=100):\n",
      "    \"\"\"Query RNACentral and yield parsed records\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    query : str\n",
      "        The query to use against RNACentral\n",
      "        \n",
      "    Returns\n",
      "    -------\n",
      "    generator\n",
      "        Yields `dict` of the individual records.\n",
      "    \n",
      "    See Also\n",
      "    --------\n",
      "    http://rnacentral.org/api\n",
      "    \"\"\"\n",
      "    def _request_to_json(url, params=None, headers=None):\n",
      "        # Perform the request, and if valid, fetch the JSON data. We're defining this as a\n",
      "        # function as we'll be performing this type of operation a few times.\n",
      "        r = requests.get(url, params=params, headers=headers)\n",
      "        if r.status_code != 200:\n",
      "            raise IOError(\"We received the following status code: %d\" % r.status_code)\n",
      "\n",
      "        return r.json()\n",
      "\n",
      "    url = 'http://rnacentral.org/api/v1/rna'\n",
      "    \n",
      "    # Tell RNAcentral that we'd like JSON data back\n",
      "    headers = {'Accept': 'application/json'}\n",
      "    \n",
      "    # These are the specific parameters for the query. For example, if our query was foo,\n",
      "    # the resulting URL would be (feel free to copy/paste into your browser too):\n",
      "    # http://rnacentral.org/api/v1/rna/?query=foo&page_size=25&flat=true\n",
      "    \n",
      "    # More information about the specific parameters and their values can be found on the\n",
      "    # API description page: http://rnacentral.org/api\n",
      "    params = {'query': query,   # The specific query\n",
      "              'page_size': 25,  # The number of items to return per page. \n",
      "              'flat': 'True'}   # This expands out the detail about each record\n",
      "    \n",
      "    data = _request_to_json(url, params, headers)\n",
      "    \n",
      "    # The key `next` contains the URL of the next \"page\" of data\n",
      "    next_page = data['next']\n",
      "    \n",
      "    # The key `count` contains the total number of items that match our query\n",
      "    count_in_payload = data['count']\n",
      "    \n",
      "    # And, let's setup a counter to track now many results have been yielded\n",
      "    count_yielded = 0\n",
      "    \n",
      "    while count_in_payload > 0 and count_yielded < max_records:\n",
      "        for record in data['results']:\n",
      "            # Let's only care about records that appear to have sequence and length data\n",
      "            sequence = record.get('sequence')\n",
      "            length = record.get('length')\n",
      "            \n",
      "            if sequence is None or length is None:\n",
      "                continue\n",
      "                \n",
      "            # The \"flat\" form from RNAcentral has a list of external references associated with each sequence.\n",
      "            # This list might contain more than a single entry if the sequence has been deposited multiple times,\n",
      "            # and for the purposes of our program, we're going to consider each one of these as independent.\n",
      "            sequence = {'sequence': sequence, 'length': length}    \n",
      "            for xref in record['xrefs']:\n",
      "                # Now, let's build our notion of a record and yield it so a consumer can operate on it\n",
      "                to_yield = sequence.copy()\n",
      "                to_yield.update(xref)\n",
      "                yield to_yield\n",
      "                count_yielded += 1\n",
      "        \n",
      "        # Finally, if there is another page of data, let's grab it, otherwise we'll terminate the loop\n",
      "        if next_page:\n",
      "            data = _request_to_json(next_page)\n",
      "        else:\n",
      "            break"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 30
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "So what do these records that we're getting from RNAcentral look like? Let's explore! For our test, let's search for \"tRNA\", and to be polite, we'll only request a single record as that's all we need right now. We're also going to use a helper method, `pprint`, to format the resulting `dict` better for human consumption."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from pprint import pprint\n",
      "test_query = query_rnacentral('tRNA', max_records=1)\n",
      "pprint(test_query.next())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "{u'accession': {u'citations': u'http://rnacentral.org/api/v1/accession/AF057362.1%3A1..1200%3ArRNA/citations',\n",
        "                u'description': u'Salmonella enterica subsp. enterica serovar Paratyphi B partial 16S ribosomal RNA',\n",
        "                u'expert_db_url': u'',\n",
        "                u'external_id': u'',\n",
        "                u'gene': u'',\n",
        "                u'id': u'AF057362.1:1..1200:rRNA',\n",
        "                u'optional_id': u'',\n",
        "                u'organelle': u'',\n",
        "                u'product': u'16S ribosomal RNA',\n",
        "                u'rna_type': u'rRNA',\n",
        "                u'source_url': u'http://www.ebi.ac.uk/ena/data/view/Non-coding:AF057362.1:1..1200:rRNA',\n",
        "                u'species': u'Salmonella enterica subsp.',\n",
        "                u'url': u'http://rnacentral.org/api/v1/accession/AF057362.1%3A1..1200%3ArRNA/info'},\n",
        " u'database': u'ENA',\n",
        " u'first_seen': u'2014-05-29',\n",
        " u'is_active': True,\n",
        " u'is_expert_db': False,\n",
        " u'last_seen': u'2014-07-25',\n",
        " 'length': 1200,\n",
        " 'sequence': u'AAAUUGAAGAGUUUGAUCCUGGCUCAGAUUGAACGCUGGCGGCAGGCCUAACACAUGCAAGUCGAACGGUAACAGGAAGCAGCUUGCUGCUUCGCUGACGAGUGGCGGACGGGUGAGUAAUGUCUGGGAAACUGCCUGAUGGAGGGGGAUAACUACUGGAAACGGUGGCUAAUACCGCAUAACGUCGCAAGACCAAAGAGGGGGACCUUCGGGCCUCUUGCCAUCAGAUGUGCCCAGAUGGGAUUAGCUAGUUGGUGAGGUAACGGCUCACCAAGGCGACGAUCCCUAGCUGGUCUGAGAGGAUGACCAGCCACACUGGAACUGAGACACGGUCCAGACUCCUACGGGAGGCAGCAGUGGGGAAUAUUGCACAAUGGGCGCAAGCCUGAUGCAGCCAUGCCGCGUGUAUGAAGAAGGCGUUCGGGUUGUAAAGUACUUUCAGCGGGGAGGAAGGUGUUGUGGUUAAUAACCGCAGCAGAUUGACGUUACCCGCAGAAGAAGCACCGGCUAACUCCGUGCCAGCAGCCGCGGUAAUACGGAGGGUGCAAGCGUUAAUCGGAAUUACUGGGCGUAAAGCGCACGCAGGCGGUCUGUCAAGUCGGAUGUGAAAUCCCCGGGCUCAACCUGGGAACUGCAUUCGAAACUGGCAGGCUUGAGUCUUGUAGAGGGGGGUAGAAUUCCAGGUGUAGCGGUGAAAUGCGUAGAGAUCUGGAGGAAUACCGGUGGGGCGAAGGCGGCCCCCUGGACAAAGACUGACGCUCAGGUGCGAAACCGUGGGGAGCAAACAGGAUUAGAUACCCUGGUAGUCCACGCCGUAAACGAUGUCUACUUGGAGGUUGUGCCCUUGAGAGCGUGGCUUCCGGAGCUAACGCGUUAAGUAGACCGCCUGGGGAGUACGGCCGCAAGCUUAAAACUCAAAUGAAUUGACGGGGGCCCGCACAAGCGGUGGAGCAUGUGGUUUAAUUCGAUCCAACGCGAAGAACCUUACCUGGUCUUGACAUCCACAGAACUUUCCAGAGAUGGACUGGUGCCUUCGGGAACUGUGAGACAGGUGCUGCAUGGCUGUCGUCAGCUCGUGUUGUGAAAUGUUGGGUUAAGUUCCGCAACGAGCGCAACCCUUAUCCUUUGUUGCCAGCGGUCCGGCCGGGAACUCAAAGGAGACUGCCAGUGAUAAACUGGAGGAAGGUGGGGAUGACGUCA',\n",
        " u'taxid': 57045}\n"
       ]
      }
     ],
     "prompt_number": 64
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Well, that's interesting. Even though we queried for \"tRNA\", we ended up getting back \"rRNA\"! Unfortunately, while the search interface of RNAcentral is quite powerful (and we're intentionally not fully leveraging it here), it is not perfect. So, we'll need to do some data scrubbing on our end if we want to only get specific records. More on this in a little. "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Next, it'd be great to get the lineage associated with the records. Luckily, RNAcentral provides a `taxid`, which corresponds to the NCBI taxonomy. Unfortunately, they don't provide the full lineage. But, we can query NCBI to gather these details. NCBI provides [EUtils](http://www.ncbi.nlm.nih.gov/books/NBK25501/) that let users programmatically query their resources similarly to RNAcentral. "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def query_ncbi_lineage(taxon_id):\n",
      "    \"\"\"Obtain the NCBI lineage for a taxon ID\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    taxon_id : int\n",
      "        The taxon ID of interest\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    list or None\n",
      "        Each taxon name or None if unable to retreive the taxon details\n",
      "    \"\"\"\n",
      "    url = \"http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
      "    \n",
      "    params = {'db': 'taxonomy',\n",
      "              'id': taxon_id}\n",
      "    \n",
      "    r = requests.get(url, params=params)\n",
      "    \n",
      "    if r.status_code != 200:\n",
      "        return None\n",
      "    \n",
      "    tree = ET.fromstring(r.content)\n",
      "    lineage = tree.iter('Lineage').next()\n",
      "    \n",
      "    if lineage is not None:\n",
      "        return [v.strip() for v in lineage.text.split(';')]\n",
      "    else:\n",
      "        return None"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 60
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "testtaxon = query_ncbi_lineage(1221535)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "failures = []\n",
      "def failed(item):\n",
      "    global failures\n",
      "    failures.append(item)\n",
      "    \n",
      "def record_parser(items, options):\n",
      "    stats = {'filtered-environmental': 0}\n",
      "    for item in items:\n",
      "        item = item.copy()\n",
      "        id_ = item['accession']['id'].split(':', 1)[0]\n",
      "        \n",
      "        if options['filter-environmental']:\n",
      "            acc = item['accession']\n",
      "            tokens = {t.lower() for t in acc['description']}\n",
      "            tokens.update({t.lower() for t in acc['species']})\n",
      "            if self.Options['filter-environmental'].intersection(tokens):\n",
      "                failed(item)\n",
      "                stats['filter-environmental'] += 1\n",
      "                continue\n",
      "                \n",
      "        item['sequence'] = item['sequence'].replace('T', 'U')\n",
      "        \n",
      "        if item['length'] < options['minimum-length']:\n",
      "            failed(item)\n",
      "            stats['minimum-length'] += 1\n",
      "            continue\n",
      "            \n",
      "        if item['length'] > options['maximum-length']:\n",
      "            failed(item)\n",
      "            stats['maximum-length'] += 1\n",
      "            continue\n",
      "            \n",
      "        if option['compute-gc']:\n",
      "            gc = float(item['sequence'].count('G') + item['sequence'].count('C'))\n",
      "            gc /= item['length']\n",
      "        else:\n",
      "            gc = None\n",
      "            \n",
      "        if option['minimum-gc'] and gc is not None:\n",
      "            if gc < options['minimum-gc']:\n",
      "                failed(item)\n",
      "                stats['minimum-gc'] += 1\n",
      "                continue\n",
      "\n",
      "        if option['maximum-gc'] and gc is not None:\n",
      "            if gc < options['maximum-gc']:\n",
      "                failed(item)\n",
      "                stats['maximum-gc'] += 1\n",
      "                continue\n",
      "        \n",
      "        if item.get('taxid', None) is not None:\n",
      "            lineage = query_ncbi_lineage(item['taxid'])\n",
      "        else:\n",
      "            lineage = None\n",
      "            stats['missing-lineage'] += 1\n",
      "            \n",
      "        if lineage is not None:\n",
      "            group = lineage[-1]\n",
      "        else:\n",
      "            group = 'unassigned'\n",
      "    \n",
      "    item['gc'] = gc\n",
      "    item['group'] = group\n",
      "    item['lineage'] = lineage\n",
      "    item['id'] = id_\n",
      "    \n",
      "    yield item"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "class RecordParser(Workflow):\n",
      "    def initialize_state(self, item):\n",
      "        self.state = item\n",
      "         \n",
      "    @method(priority=99)\n",
      "    @requires(option='filter-environmental')\n",
      "    def filter_environmental(self):\n",
      "        acc = self.state['accession']\n",
      "        tokens = {t.lower() for t in acc['description']}\n",
      "        tokens.update({t.lower() for t in acc['species']})\n",
      "        if self.Options['filter-environmental'].intersection(tokens):\n",
      "            self.failed = True\n",
      "    \n",
      "    @method(priority=95)\n",
      "    def force_to_rna(self):\n",
      "        self.state['sequence'] = self.state['sequence'].replace('T', 'U')\n",
      "    \n",
      "    @method(priority=89)\n",
      "    @requires(option='minimum-length')\n",
      "    def minimum_length(self):\n",
      "        if self.state['length'] < self.Options['minimum-length']:\n",
      "            self.failed = True\n",
      "    \n",
      "    @method(priority=88)\n",
      "    @requires(option='maximum-length',  state=has_length)\n",
      "    def maximum_length(self):\n",
      "        if self.state['length'] > self.Options['maximum-length']:\n",
      "            self.failed = True\n",
      "    \n",
      "    @method(priority=80)\n",
      "    @requires(option='compute-gc')\n",
      "    def compute_gc(self):\n",
      "        gc = self.state['sequence'].count('G') + self.state['sequence'].count('C')\n",
      "        self.state['gc'] = float(gc) / self.state['length']\n",
      "    \n",
      "    @method(priority=79)\n",
      "    @requires(option='minimum-gc', state=has_gc)\n",
      "    def minimum_gc(self):\n",
      "        if self.state['gc'] < self.Options['minimim-gc']:\n",
      "            self.failed = True\n",
      "    \n",
      "    @method(priority=78)\n",
      "    @requires(option='maximum-gc', state=has_gc)\n",
      "    def maximum_gc(self):\n",
      "        if self.state['gc'] > self.Options['maximum-gc']:\n",
      "            self.failed = True\n",
      "    \n",
      "    @method(priority=2)\n",
      "    @requires(state=has_taxonid)\n",
      "    def fetch_lineage():\n",
      "        # http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=taxonomy&id=1221535\n",
      "        pass\n",
      "    \n",
      "    @method(priority=1)\n",
      "    def assign_to_group(self):\n",
      "        # fail if we don't have lineage\n",
      "        pass\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "NameError",
       "evalue": "name 'Workflow' is not defined",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
        "\u001b[0;32m<ipython-input-20-1c0913fc7e52>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mclass\u001b[0m \u001b[0mRecordParser\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mWorkflow\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0minitialize_state\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstate\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mitem\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpriority\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;31mNameError\u001b[0m: name 'Workflow' is not defined"
       ]
      }
     ],
     "prompt_number": 20
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "class RecordA"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}